bftengine/src/preprocessor/PreProcessor.cpp

```diff
@@ -207,18 +207,39 @@ void RequestsBatch::releaseReqsAndSendBatchedReplyIfCompleted(PreProcessReplyMsg
 }
 
 // On a primary replica
+void RequestsBatch::cancelRequestAndBatchIfCompleted(const string &reqBatchCid,
+                                                     uint16_t reqOffsetInBatch,
+                                                     PreProcessingResult status) {
+  const lock_guard<mutex> lock(batchMutex_);
+  if (batchCid_ != reqBatchCid) {
+    LOG_INFO(preProcessor_.logger(),
+             "The batch has been cancelled/completed earlier; do nothing" << KVLOG(clientId_, reqBatchCid, batchCid_));
+    return;
+  }
+  const auto &reqEntry = requestsMap_[reqOffsetInBatch];
+  if (reqEntry) {
+    if (status == CANCEL) preProcessor_.preProcessorMetrics_.preProcConsensusNotReached++;
+    preProcessor_.releaseClientPreProcessRequestSafe(clientId_, reqEntry, status);
+    finalizeBatchIfCompleted();
+  }
+}
+
+// On a primary replica
+void RequestsBatch::finalizeBatchIfCompletedSafe() {
+  const lock_guard<mutex> lock(batchMutex_);
+  finalizeBatchIfCompleted();
+}
+
+// On a primary replica; unsafe
 void RequestsBatch::finalizeBatchIfCompleted() {
   string batchCid;
   atomic_uint32_t batchSize = 0;
   uint32_t numOfCompletedReqs = 0;
-  {
-    const lock_guard<mutex> lock(batchMutex_);
-    if (numOfCompletedReqs_ < batchSize_) return;
-    batchCid = batchCid_;
-    batchSize = batchSize_;
-    numOfCompletedReqs = numOfCompletedReqs_;
-    resetBatchParams();
-  }
+  if (numOfCompletedReqs_ < batchSize_) return;
```

Comment:
if we in the cancellation flow, and ```numOfCompletedReqs_ < batchSize_```
What is the future on this batch?

Reply:
Does not matter how the request completes - successfully or has been canceled, numOfCompletedReqs_ increases.